### DISCUSSION TITLE
Make auth JWTs http-only cookies (by default)

### DISCUSSION THREAD

**Discussion by Arkenar** [supabase/supabase#12303] [Feature Requests]
Correct me if i am mistaken but the default way of sending cookies to the client is using local storage. If i want to setup http-only cookies, i have to add another webserver as a middleware that would take the Supabase response and replace the header being sent with an http-only cookie. 

This is a major concern for me. I tend to use JWT tokens as data carriers for ABAC authorization on my services (the signing key is first verified then a service-scoped policy uses the JWT claims to handle authZ locally and without making a call to an IAM server or to a gateway) and i don't want this data to be exposed to the client.

Is there a way to send http-only cookies directly from Supabase? If not, could you please add that feature to the app as soon as possible? I could add a middleware webserver but that would kill the purpose of using Supabase for authN. When it comes to session security, providing at least an option to configure http-only tokens which cannot be access from client scripts is a must, especially when many Supabase users are entry level developers. The lack of this feature might expose a large number of applications to vulnerabilities (mostly to XSS attacks but not limited to that).

**Comment by dbmikus** [comment]
I'd like to +1 this, although I don't need http-only cookie auth by default. I've needed cookie-based auth before and Supabase doesn't provide much easy support for that, last time I checked.

I need my session cookies to come from my API server (same server I host my webapp on) and not the Supabase GoTrue server. Perhaps on my API server I could make requests to GoTrue to create and check session info and then put it in a cookie to send to the client? It would be helpful if Supabase streamlined this.

I did end up creating my own session cookie system, but at that point I feel like I'm fighting against Supabase.

**Comment by omarelshiltawi** [comment]
let me know if this helps https://chat.elapse.ai/conversation/7da14426-e0cd-40ac-9485-709ccacc55bd

**Comment by sam4work** [comment]
I also noticed this and when I checked the docs,  it's state that it is not need. I wonder how they handle the security issues relating to setting cookie httpOnly  option to false 

https://supabase.com/docs/guides/auth/server-side-rendering#how-do-i-make-the-cookies-httponly-

**Comment by KrisBraun** [comment]
I think I need HTTP-only cookies to work around the size limit (see [this issue](https://github.com/supabase/gotrue/issues/1160) for details). Does anyone have any pointers to setting that up? @omarelshiltawi your link now 404s.

**Comment by maddy020** [comment]
I am using supabase for authentication with nextjs  and writing my  backend API  in express 
Cookies are being set in http server(my localhost) but when set over production(https) cookies are not passed with the request headers
How can I change the configuration of cookie set by supabase
I have checked in my application tab , secure attribute is false
May be I changed this to true and the things get in work
Correct me If I am wrong..

But the  main question is how to chang the configuration?So that I can access the token set in the cookies in express to verify a middleware

**Comment by hf** [comment]
HTTP-Only cookies are not an option _unless you're making a very traditional web app_ using frameworks like Rails, Laravel, Django, traditional Express etc. Even then, use of JavaScript on the frontend should be limited and not be interactive (i.e. no third-party API calls to Supabase Data APIs for example).

If that's the case, you must use Supabase Auth on the server only and configure the `storage` option of the Supabase JavaScript client (for a traditional Node.js Express app) to use the request and response cookies while setting the HTTP-Only option.

**Comment by j4w8n** [comment]
Yeah, likely never gonna be the default in Supabase.

I recently played around with this, and it's easy to do a basic implementation using Supabase's SSR package: https://discord.com/channels/839993398554656828/1240082081183436830/1240082081183436830

**Reply by frankdagostino** [reply]
the link doesn't work! could you give me an example here please? 

I need an implementation of httponly cookies... **tokens on the browser side is terrible!**

**Reply by j4w8n** [reply]
The link worked for me, but below is the content. To be more specific, the key code in the layout.ts file is the browser client's cookie implementation. And as for the hooks.server code, it'd be the supabase client's cookie options and the /cookies route that I implement further down the file.


"I had a thought earlier, that it might not be that bad to implement httpOnly cookies with the ssr package. I got it working on my demo app, but would like to get some other eyes on it, to see if there are any flaws here.

There are going to be caveats, but I'm curious about the basic implementation right now.

src/routes/+layout.ts and /src/hooks.server.ts are going to be the key files.

Using a /cookies route to return the session."

https://github.com/j4w8n/sveltekit-supabase-ssr/tree/httponly

**Reply by timothyerwin** [reply]
isn't this still inherently insecure if you can access the session anyway via the route? it seems like one indirection removed but could still be visible on the client side which is the whole point httponly enforces

**Reply by j4w8n** [reply]
> isn't this still inherently insecure if you can access the session anyway via the route? it seems like one indirection removed but could still be visible on the client side which is the whole point httponly enforces

Yeah, wow. I hadn't thought of that, for whatever reason. If the point of using httponly is to keep client-side js from accessing cookies, then having a wide open endpoint that client-side js can query isn't doing anything. 🤦‍♂️

**Comment by tysonzero** [comment]
@hf I don't see why HttpOnly cookies wouldn't be an option, as long as some sort of supabase-auth sdk middleware is running on the servers you are communicating with, you should be perfectly able to do all the usual non-HTML AJAX and SPA stuff just fine, and the middleware will take care of validating the access_token JWT and refreshing the refresh_token whenever the former is expiring or about to expire. For SSR to work isn't that sort of fully-cookie-based flow needed anyway?

**Comment by searles9** [comment]
**Is there no solution to this?**

I followed the docs: https://supabase.com/docs/guides/auth/server-side/nextjs?queryGroups=router&router=app

The middleware is giving the client a cookie, but its not an httpOnly cookie. 
I see no information on how to make that cookie an http only cookie. 

This is the code that the docs show I should use (which doesnt give an http only cookie):
```
import { createServerClient } from '@supabase/ssr'
import { NextResponse, type NextRequest } from 'next/server'

export async function updateSession(request: NextRequest) {
  let supabaseResponse = NextResponse.next({
    request,
  })

  const supabase = createServerClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
    {
      cookies: {
        getAll() {
          return request.cookies.getAll()
        },
        setAll(cookiesToSet) {
          cookiesToSet.forEach(({ name, value, options }) => request.cookies.set(name, value))
          supabaseResponse = NextResponse.next({
            request,
          })
          cookiesToSet.forEach(({ name, value, options }) =>
            supabaseResponse.cookies.set(name, value, options)
          )
        },
      },
    }
  )

  // Do not run code between createServerClient and
  // supabase.auth.getUser(). A simple mistake could make it very hard to debug
  // issues with users being randomly logged out.

  // IMPORTANT: DO NOT REMOVE auth.getUser()

  const {
    data: { user },
  } = await supabase.auth.getUser()

  if (
    !user &&
    !request.nextUrl.pathname.startsWith('/login') &&
    !request.nextUrl.pathname.startsWith('/auth')
  ) {
    // no user, potentially respond by redirecting the user to the login page
    const url = request.nextUrl.clone()
    url.pathname = '/login'
    return NextResponse.redirect(url)
  }

  // IMPORTANT: You *must* return the supabaseResponse object as it is.
  // If you're creating a new response object with NextResponse.next() make sure to:
  // 1. Pass the request in it, like so:
  //    const myNewResponse = NextResponse.next({ request })
  // 2. Copy over the cookies, like so:
  //    myNewResponse.cookies.setAll(supabaseResponse.cookies.getAll())
  // 3. Change the myNewResponse object to fit your needs, but avoid changing
  //    the cookies!
  // 4. Finally:
  //    return myNewResponse
  // If this is not done, you may be causing the browser and server to go out
  // of sync and terminate the user's session prematurely!

  return supabaseResponse
}
```

**Reply by danfranco3** [reply]
Do you have any answer on this? I'm experiencing the same issue, but the docs apparently say that we do not need to set the cookie options explicitly. Maybe I'll have to refactor the code to only call supabase server-side.

**Reply by searles9** [reply]
@danfranco3 Yeah look at my code below. I was able to get it to give an HTTP only cookie:

```
import { createServerClient } from '@supabase/ssr'
import { NextResponse, type NextRequest } from 'next/server'

export async function updateSession(request: NextRequest) {
  let supabaseResponse = NextResponse.next({
    request,
  })

  const supabase = createServerClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
    {
      cookies: {
        getAll() {
          return request.cookies.getAll()
        },
        setAll(cookiesToSet) {
          // This is necessary because the middleware needs to update the request object 
          // with the new cookie values before creating the response. 
          // This ensures that any subsequent middleware or route handlers 
          // have access to the updated cookie values.
          // TLDR: This makes sure the request object has the updated cookie values for server-side operations
          cookiesToSet.forEach(({ name, value, options }) => request.cookies.set(name, value))
          supabaseResponse = NextResponse.next({
            request,
          })
          
          // This sets the cookies on the response object that will be sent back to the client.
          // The additional options (httpOnly, secure, etc.) are important security settings 
          // that should be applied to the cookies sent to the browser.
          // TLDR: This makes sure the client gets the cookies
          cookiesToSet.forEach(({ name, value, options }) =>
            supabaseResponse.cookies.set(name, value, {
              ...options,       // <-- existing options from Supabase (may include expires, etc)
              httpOnly: true,   // <-- blocks JavaScript access (XSS protection)
              secure: process.env.NODE_ENV === "production", // <-- blocks cookies on non-HTTPS in production
              sameSite: "lax",  // <-- prevents most CSRF attacks but allows normal navigation
              path: "/",        // <-- ensures the cookie is sent on all routes
            })
          )
        },
      },
    }
  )

  // Do not run code between createServerClient and
  // supabase.auth.getUser(). A simple mistake could make it very hard to debug
  // issues with users being randomly logged out.

  // IMPORTANT: DO NOT REMOVE auth.getUser()

  const {
    data: { user },
  } = await supabase.auth.getUser()

  // Define public routes 
  const isPublicRoute = 
    request.nextUrl.pathname === '/' ||
    request.nextUrl.pathname === '/pricing' ||
    request.nextUrl.pathname === '/terms' ||
    request.nextUrl.pathname === '/privacy' ||
    request.nextUrl.pathname.startsWith('/auth')

  // Case 1: If the user IS logged in and trying to access a public route, redirect to the accounts page
  if (user && isPublicRoute) {
    const url = request.nextUrl.clone()
    // TO-DO: where should I redirect them to if they are logged in? We no longer have an accounts page 
    url.pathname = '/accounts'
    return NextResponse.redirect(url)
  }

  // Case 2: If the user IS logged in and trying to access a private route, do nothing
  // (This case is handled by default - no redirect needed)

  // Case 3: If the user is NOT logged in and trying to access a private route redirect to login
  if (!user && !isPublicRoute) {
    const url = request.nextUrl.clone()
    url.pathname = '/auth/login'
    return NextResponse.redirect(url)
  }

  // IMPORTANT: You *must* return the supabaseResponse object as it is.
  // If you're creating a new response object with NextResponse.next() make sure to:
  // 1. Pass the request in it, like so:
  //    const myNewResponse = NextResponse.next({ request })
  // 2. Copy over the cookies, like so:
  //    myNewResponse.cookies.setAll(supabaseResponse.cookies.getAll())
  // 3. Change the myNewResponse object to fit your needs, but avoid changing
  //    the cookies!
  // 4. Finally:
  //    return myNewResponse
  // If this is not done, you may be causing the browser and server to go out
  // of sync and terminate the user's session prematurely!

  return supabaseResponse
}
```

**Comment by zakzackr** [comment]
I have recently built a Next.js demo showcasing Supabase SSR authentication with httpOnly cookies. 

This approach improves security by preventing XSS attacks, but requires workarounds since `onAuthStateChange()` and `getUser()/getClaims()` cannot be used on the client-side due to httpOnly cookie restrictions.

Demo Repository: https://github.com/zakzackr/supabase-ssr-httponly-demo
